usaon_vta_survey/routes/response/relationships/observing_system_data_product.py

```diff
@@ -0,0 +1,222 @@
+            ResponseObservingSystemDataProduct,
+            (data_product_id, observing_system_id),
+        )
+    else:
+        response_observing_system_data_product = None
+
+    if response_observing_system_data_product is not None:
+        return response_observing_system_data_product
+    else:
+        return ResponseObservingSystemDataProduct()
+
+
+def _request_args(request: Request) -> tuple[int | None, int | None]:
+    data_product_id: int | str | None = request.args.get('data_product_id')
+    if data_product_id is not None:
+        data_product_id = int(data_product_id)
+
+    observing_system_id: int | str | None = request.args.get('observing_system_id')
+    if observing_system_id is not None:
+        observing_system_id = int(observing_system_id)
+
+    return data_product_id, observing_system_id
+
+
+@app.route(
+    '/response/<string:survey_id>/observing_system_data_product_relationships',
+    methods=['GET', 'POST'],
+)
+def view_response_observing_system_data_product_relationships(survey_id: str):
+    """View and add observing system/dataproduct relationships to a response.
+
+    TODO: Refactor this whole pile of stuff. Less string magic. Less cyclomatic
+    complexity.
+    """
+    data_product_id, observing_system_id = _request_args(request)
+    survey = db.get_or_404(Survey, survey_id)
+
+    class SuperForm(FlaskForm):
+        """Combine all necessary forms into one super-form.
+
+        NOTE: Additional class attributes are added dynamically below.
+        """
+
+        relationship = FormField(FORMS_BY_MODEL[ResponseObservingSystemDataProduct])
+
+    response_observing_system_data_product = _response_observing_system_data_product(
+        data_product_id=data_product_id,
+        observing_system_id=observing_system_id,
+    )
+
+    response_data_product = _response_data_product(
+        data_product_id=data_product_id,
+        response_id=survey.response_id,
+    )
+
+    response_observing_system = _response_observing_system(
+        observing_system_id=observing_system_id,
+        response_id=survey.response_id,
+    )
+
+    _update_super_form(
+        SuperForm,
+        observing_system_id=observing_system_id,
+        data_product_id=data_product_id,
+    )
+    _update_relationship(
+        response_observing_system_data_product,
+        observing_system_id=observing_system_id,
+        data_product_id=data_product_id,
+    )
+
+    form_obj: dict[
+        str,
+        ResponseObservingSystem
+        | ResponseDataProduct
+        | ResponseObservingSystemDataProduct,
+    ] = {
+        'observing_system': response_observing_system,
+        'data_product': response_data_product,
+        # NOTE: Logic below depends on relationship being last in this dict
+        'relationship': response_observing_system_data_product,
+    }
+
+    if request.method == 'POST':
+        limit_response_editors()
+        form = SuperForm(request.form, obj=form_obj)
```

Comment:
Was this "super form" stuff intuitive enough?

Reply:
yes!